packages/navigator/src/browser/navigator-contribution.ts

```diff
@@ -358,6 +382,20 @@ export class FileNavigatorContribution extends AbstractViewContribution<FileNavi
                 });
             }
         });
+        registry.registerCommand(OpenEditorsCommands.CLOSE_ALL_TABS_FROM_TOOLBAR, {
+            execute: widget => this.withOpenEditorsWidget(widget, () => this.editorWidgets.forEach(editor => editor.close())),
+            isEnabled: widget => this.withOpenEditorsWidget(widget, () => !!this.editorWidgets.length),
+            isVisible: widget => this.withOpenEditorsWidget(widget, () => !!this.editorWidgets.length)
+        });
+        registry.registerCommand(OpenEditorsCommands.SAVE_ALL_TABS_FROM_TOOLBAR, {
+            execute: widget => this.withOpenEditorsWidget(widget, () => this.commandService.executeCommand(CommonCommands.SAVE_ALL.id)),
+            isEnabled: widget => this.withOpenEditorsWidget(widget, () => !!this.editorWidgets.length),
+            isVisible: widget => this.withOpenEditorsWidget(widget, () => !!this.editorWidgets.length)
+        });
+    }
+
+    protected get editorWidgets(): NavigatableWidget[] {
```

Comment:
Rather than putting this here, could you have a field on the `OpenEditorsWidget` itself that tracks the number of widgets it contains and access that, instead? It would save a couple of filter calls, at least. It should be `get`able as something like `this.model.root?.children?.length ?? 0`, though I'm sure it'll want a `CompositeTreeNode.is()` check rather than just pretending that `children` might exist on *any* old tree node!

Reply:
I've kept the getter here for convenience but exposed a getter through the tree model that retrieves the editorWidgets. I also had to use the `widgetManager` (rather than `this.widget`) to access the `OpenEditorsWidget` since the `navigator-contribution`'s widget is actually the `FileNavigatorWidget`